xrdp/xrdp_mm.c

```diff
@@ -1192,24 +1094,233 @@ dynamic_monitor_data(intptr_t id, int chan_id, char *data, int bytes)
     in_uint32_le(s, monitor_layout_size);
     if (monitor_layout_size != 40)
     {
+        /* 2.2.2.2 DISPLAYCONTROL_MONITOR_LAYOUT_PDU */
         LOG(LOG_LEVEL_ERROR, "dynamic_monitor_data: monitor_layout_size"
             " is %d. Per spec ([MS-RDPEDISP] 2.2.2.2"
             " DISPLAYCONTROL_MONITOR_LAYOUT_PDU) it must be 40.",
             monitor_layout_size);
         return 1;
     }
 
-    struct display_size_description *description =
-        (struct display_size_description *)
-        g_malloc(sizeof(struct display_size_description), 1);
+    dynamic_description = (struct dynamic_monitor_description *)
+                          g_malloc(sizeof(struct dynamic_monitor_description), 1);
+
+    error = libxrdp_process_monitor_stream(s, &(dynamic_description->description), 1);
+    list_add_item(wm->mm->resize_queue, (tintptr)dynamic_description);
+    dynamic_description->start_time = g_time3();
+    advance_resize_state_machine(wm->mm->resize_state_machine, dynamic_description, WMRZ_QUEUED);
+    return 0;
+}
+
+const char *
+wm_login_state_to_str(enum wm_login_state login_state);
```

Comment:
A function prototype not in a header is rarely a good idea. You could rename this function as `xrdp_wm_login_state_to_str()` and add the prototype to xrdp/xrdp.h along with all the other wm functions.

A better idea might be to replace this code which occurs twice with minor variations:-

```c
    if (wm->login_state != WMLS_CLEANUP && wm->login_state != WMLS_INACTIVE)
    {
        LOG(LOG_LEVEL_INFO, "process_dynamic_monitor_description: Not trying to resize while logging in! State was %s", wm_login_state_to_str(wm->login_state));
        return 0;
    }
```

with a call to a function like `xrdp_wm_can_resize()`:-

```c
int
xrdp_wm_can_resize(struct xrdp_wm *self)
{
    int rv = (self->login_state != WMLS_CLEANUP && self->login_state != WMLS_INACTIVE);
    if (!rv)
    {
        LOG(LOG_LEVEL_INFO, "Not trying to resize while logging in!");
        LOG_DEVEL(LOG_LEVEL_INFO, "State is %s", wm_login_state_to_str(self->login_state));
    }
    return rv;
}
```

That better keeps the encapsulation of the `login_state` member. The above is off the top of my head, but I'm sure you get the idea. Prototype for `xrdp_wm_can_resize()` lives in xrdp/xrdp.h.

Reply:
Done.